### ISSUE TITLE
How to backup all streams of a NATS Streaming Server?

### ISSUE THREAD

**Issue by hossainemruz** [nats-io/natscli#213]
**Problem:**

I have to backup all the streams of my NATS streaming server. I don't know what streams are there. I want to backup all the streams into a single snapshot. Currently, we can backup only a single stream using the following command.

```bash
$ nats stream backup <stream-name> /backup/dir/<stream-name>.tgz
```

**What I have tried so far:**

I have tried providing a wildcard instead of `<stream-name>`. It does not work.

```bash
$ nats stream backup * /backup/dir/backup.tgz
nats: error: "*" is not a valid stream name, try --help
```

**Possible Workaround:**

At first, I can list all the streams using `nats str ls` command. Then, I can loop through all the streams and backup them individualy.

However, this does not satisfy my requirement as I want to backup all the streams into a single snapshot. My snapshot should represent a complete state of the NATS streaming server just not a single stream.

**Comment by ripienaar** [comment]
We do not have a way to do that, each stream goes into back up state independently, message ingest does not pause during snapshots so you'd never get a exact point in time snapshot.

You can shut one of the replicas assuming all is R3 or greater and back up its disk - I know not ideal.

**Comment by hossainemruz** [comment]
Thank you @ripienaar for your response. Do you have any plan to support the workaround? For example, if the stream name is a wildcard, then the backup command will automatically find the respective streams and backup them. Probably, in a single `*.tgz` file but one file for each stream is also fine.

**Comment by ripienaar** [comment]
```
for stream in $(nats s ls -j|jq -r '.[]')
do
  nats stream backup ....
done
```

**Comment by hossainemruz** [comment]
Yes. That's what I am planning for now. However, it would have been great if nats has built-in support for this use-case. We can contribute the feature if you are willing to accept.

**Comment by ripienaar** [comment]
yes, should be a easy thing to add to the stream backup command, I'd then take the supplied directory and do `dir/<stream name>/backup files...` so you get a series of stream backups in one dir.  Today each backup has to be in a dir of its own

**Comment by hossainemruz** [comment]
Let us know if you have bandwidth to do it or need any help from us. It's not an emergency.

Btw, thank you for the amazing works.

**Comment by ripienaar** [comment]
We'd love a contribution if you want to give it a go

**Comment by hossainemruz** [comment]
Sure. Will do ASAP.

**Comment by joelhoisko** [comment]
Maybe an offtopic comment, but just so that I'm not going insane: this issue is about backing up streams in JetStream and **not** NATS Streaming Server (the precursor to JetStream) right? I'm asking, because I've been searching for information about backups/restores for NATS Streaming Server specifically and to my understanding the NATS Streaming Server doesn't have a concept of streams nor does it really work with the `nats` CLI tool either. Or am I missing something here?

**Comment by ripienaar** [comment]
This issue relates to JetStream, not Streaming Server indeed.
